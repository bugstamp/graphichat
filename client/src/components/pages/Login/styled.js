import { fadeInUp, fadeIn, headShake } from 'react-animations';
import styled, { keyframes } from 'styled-components';
import { backgrounds } from 'polished';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

import blue from '@material-ui/core/colors/blue';

import { getStyledProps, getSpacing } from '../../../styles';
import bgImage from '../../../assets/images/login-bg__1920_65.jpg';

const fadeInUpAnimation = keyframes`${fadeInUp}`;
const fadeInAnimation = keyframes`${fadeIn}`;
const headShakeAnimation = keyframes`${headShake}`;

export const LoginWrapper = styled(Grid)`
  flex: 1 auto;
  display: flex;
  position: relative;
  ${backgrounds(`url(${bgImage})`, 'no-repeat')};
  background-size: cover;
  background-position: center;
`;

export const BrandTitleWrapper = styled.div`
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  margin-bottom: ${getSpacing(2)};
  animation: 1s ${fadeInUpAnimation};
  animation-delay: .5s;
  animation-fill-mode: forwards;
  opacity: 0;

  p {
    margin-left: ${getSpacing(1)};
    font-weight: bold;
    background: linear-gradient(to right, ${blue[100]} 0%, ${blue[500]} 100%);
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
  }
`;


export const LoginPresentationWrapper = styled.div`
  flex: 1 auto;
  display: flex;
  flex-flow: column;
  align-items: center;
  justify-content: center;
`;

export const SubTitle = styled.div`
  animation: 1s ${fadeInAnimation};
  animation-delay: 1s;
  animation-fill-mode: forwards;
  color: #fff;
  padding: 0 ${getSpacing(1)};
  text-align: center;
  opacity: 0;

  p {
    font-weight: bold;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const smDown = breakpoints.down('sm');
    const fontSize = getStyledProps('theme.typography.h5.fontSize')(props);

    return `
      ${smDown} {
        font-size: ${fontSize};
      }
    `;
  }};
  }

  button {
    display: inline-block;
    position: relative;
    border: none;
    outline: none;
    color: inherit;
    background-color: transparent;
    cursor: pointer;
    z-index: 0;
    -webkit-transform: translate3d(0,0,0);
    animation-name: ${({ stopAnimation }) => (!stopAnimation && headShakeAnimation)};
    animation-duration: 1.5s;
    animation-delay: 2.5s;
    animation-iteration-count: infinite;

    &:hover {
      color: ${getStyledProps('theme.palette.primary.main')}
    }

    p {
      display: inherit;
    }
  }
`;

export const LoginFormWrapper = styled.div`
  && {
    width: 100%;
    height: 100%;
    max-height: 100%;
    max-width: 375px;
    min-width: 320px;
    position: relative;
    padding: ${getSpacing(5)} ${getSpacing(3)};
    overflow: hidden auto;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const spacing = getStyledProps('theme.spacing')(props);
    const xsDown = breakpoints.down('xs');

    return `
      ${xsDown} {
        max-width: 100%;
        padding: ${spacing(2)}px;
      }
    `;
  }}
  }
`;

export const Header = styled(Typography)`
  width: 100%;
  position: relative;
  text-align: center;
`;

export const SignUpButton = styled(Button)`
  && {
    margin-top: 1em;
  }
`;
