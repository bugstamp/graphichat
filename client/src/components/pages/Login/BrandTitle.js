import React from 'react';

import Typography from '@material-ui/core/Typography';
import useMediaQuery from '@material-ui/core/useMediaQuery';

import Logo from '../../common/Logo';
import { BrandTitleWrapper } from './styled';

const getLogoParams = (downMd, downSm) => {
  if (downSm) {
    return [35, 'h3'];
  }
  if (downMd) {
    return [35, 'h2'];
  }
  return [60, 'h1'];
};

const BrandTitle = () => {
  const downMd = useMediaQuery(theme => theme.breakpoints.down('md'));
  const downSm = useMediaQuery(theme => theme.breakpoints.down('sm'));
  const [logoSize, typographyVariant] = getLogoParams(downMd, downSm);

  return (
    <BrandTitleWrapper>
      <Logo size={logoSize} />
      <Typography
        component="p"
        variant={typographyVariant}
        align="center"
      >
        GraphiChat
      </Typography>
    </BrandTitleWrapper>
  );
};

export default BrandTitle;
