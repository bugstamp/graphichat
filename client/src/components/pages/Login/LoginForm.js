import React from 'react';
import PropTypes from 'prop-types';

import AccountCircleIcon from '@material-ui/icons/AccountCircleRounded';

import Form from '../../common/Form/Form';
import SocialMedia from '../../common/SocialMedia/SocialMedia';
import TopProgressLine from '../../common/TopProgressLine';

import { LoginFormWrapper, Header, SignUpButton } from './styled';
import { mutationResultProps } from '../../propTypes';

const LoginForm = ({
  signIn,
  signInResult,
  signInBySocial,
  signInBySocialResult,
  redirectToSignUp,
}) => {
  const { loading } = signInResult;

  return (
    <LoginFormWrapper>
      <TopProgressLine loading={loading} />
      <Header variant="h1" color="primary" align="center" gutterBottom>
        <AccountCircleIcon fontSize="inherit" color="primary" />
      </Header>
      <Form
        formId="signIn"
        mutation={signIn}
        result={signInResult}
        submitButtonText="Sign In"
      />
      <SignUpButton
        onClick={redirectToSignUp}
        color="primary"
        size="large"
        variant="outlined"
        fullWidth
      >
        Sign Up
      </SignUpButton>
      <SocialMedia
        mutation={signInBySocial}
        result={signInBySocialResult}
      />
    </LoginFormWrapper>
  );
};

LoginForm.propTypes = {
  signIn: PropTypes.func.isRequired,
  signInResult: PropTypes.shape(mutationResultProps).isRequired,
  signInBySocial: PropTypes.func.isRequired,
  signInBySocialResult: PropTypes.shape(mutationResultProps).isRequired,
  redirectToSignUp: PropTypes.func.isRequired,
};

export default LoginForm;
