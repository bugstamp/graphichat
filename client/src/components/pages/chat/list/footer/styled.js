import styled from 'styled-components';

import { getSpacing } from '../../../../../styles';

export const ChatListFooterStyled = styled.div`
  display: flex;
  justify-content: center;
  padding: ${getSpacing(2)} ${getSpacing(1)};

  && {
    button {
      width: 50%;
    }
  }
`;

export default null;
