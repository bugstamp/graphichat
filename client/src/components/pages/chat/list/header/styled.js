import styled from 'styled-components';
import { position } from 'polished';

import IconButton from '@material-ui/core/IconButton';

import { getSpacing } from '../../../../../styles';

export const ChatListHeaderStyled = styled.div`
  display: flex;
  flex-flow: column;
  padding: ${getSpacing(2)} ${getSpacing(1)};
`;

export const HeaderTitle = styled.div`
  position: relative;
  margin-bottom: ${getSpacing(1)};
`;

export const AddButton = styled(IconButton)`
  && {
    ${position('absolute', 0, 0, null, null)};
    padding: ${getSpacing(0.5)};
  }
`;
