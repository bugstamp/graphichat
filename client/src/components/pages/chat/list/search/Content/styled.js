import styled from 'styled-components';

import DialogContent from '@material-ui/core/DialogContent';

import { SearchBoxStyled } from '../../../../../common/SearchBox/styled';
import { getStyledProps, getSpacing } from '../../../../../../styles';

export const SearchDialogContentStyled = styled(DialogContent)`
  && {
    flex: 1 auto;
    display: flex;
    flex-flow: column;
    padding: ${getSpacing(2)} ${getSpacing(3)};
    padding-bottom: 0;

    ${SearchBoxStyled} {
      margin-bottom: ${getSpacing(2)}
    }

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const spacing = getStyledProps('theme.spacing')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${SearchBoxStyled} {
        margin-bottom: ${getSpacing(1)}
      }
      ${smDown} {
        padding: ${spacing(1)}px ${spacing(2)}px;
      }
    `;
  }}
  }
`;

export default null;
