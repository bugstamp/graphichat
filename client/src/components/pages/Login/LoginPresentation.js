import React, { memo } from 'react';
import PropTypes from 'prop-types';

import Typography from '@material-ui/core/Typography';

import BrandTitle from './BrandTitle';
import { LoginPresentationWrapper, SubTitle } from './styled';

const LoginPresentation = ({ stopAnimation, toggleForm }) => (
  <LoginPresentationWrapper>
    <BrandTitle />
    <SubTitle stopAnimation={stopAnimation}>
      <Typography component="p" variant="h4" align="center" paragraph>
        A lightweight, simple and useful web chat app based on the modern GraphQL API
      </Typography>
      <button type="button" onClick={toggleForm}>
        <Typography component="p" variant="h4" align="center">Try it now!</Typography>
      </button>
    </SubTitle>
  </LoginPresentationWrapper>
);

LoginPresentation.propTypes = {
  stopAnimation: PropTypes.bool.isRequired,
  toggleForm: PropTypes.func.isRequired,
};

export default memo(LoginPresentation);
