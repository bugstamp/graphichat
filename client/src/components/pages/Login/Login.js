import React from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation } from '@apollo/client';

import Drawer from '@material-ui/core/Drawer';
import useMediaQuery from '@material-ui/core/useMediaQuery';

import LoginForm from './LoginForm';
import LoginPresentation from './LoginPresentation';
import FullSwipeableDrawerStyled from '../../common/FullWidthSwipeableDrawer';
import withNotification from '../../common/HOC/withNotification';

import storage from '../../../storage';
import history from '../../../router/history';
import gql from '../../../gql';

import { LoginWrapper } from './styled';

const {
  CHECK_SESSION_EXPIRATION,
  SIGN_IN,
  SIGN_IN_BY_SOCIAL,
} = gql;

const Login = (props) => {
  const { toggleNotification } = props;
  const [form, toggleForm] = React.useState(false);
  const { data: { sessionExpired } } = useQuery(CHECK_SESSION_EXPIRATION);
  const smUp = useMediaQuery(theme => theme.breakpoints.up('sm'));
  const DrawerComponent = smUp
    ? Drawer
    : FullSwipeableDrawerStyled;

  React.useEffect(() => {
    if (sessionExpired) {
      toggleNotification('Session time was expired');
    }
  }, [sessionExpired, toggleNotification]);

  const handleToggleForm = React.useCallback(() => {
    toggleForm(!form);
  }, [form]);

  function redirectToSignUp() {
    history.push('/reg');
  }

  function handleSuccess({ token, refreshToken }) {
    storage.setTokens(token, refreshToken);
    history.push('/');
  }

  function handleError(e) {
    if (e.graphQLErrors) {
      const { graphQLErrors } = e;
      const { message, data = null } = graphQLErrors[0];

      if (data) {
        const { invalidField = null } = data;

        if (!invalidField) {
          toggleNotification(message);
        }
      }
    }
  }

  const [signIn, signInResult] = useMutation(SIGN_IN, {
    onCompleted: data => handleSuccess(data.signIn),
    onError: handleError,
  });
  const [signInBySocial, signInBySocialResult] = useMutation(SIGN_IN_BY_SOCIAL, {
    onCompleted: data => handleSuccess(data.signInBySocial),
    onError: handleError,
  });

  return (
    <LoginWrapper container>
      <LoginPresentation stopAnimation={form} toggleForm={handleToggleForm} />
      <DrawerComponent
        open={form}
        onClose={handleToggleForm}
        anchor="right"
      >
        <LoginForm
          signIn={signIn}
          signInResult={signInResult}
          signInBySocial={signInBySocial}
          signInBySocialResult={signInBySocialResult}
          redirectToSignUp={redirectToSignUp}
        />
      </DrawerComponent>
    </LoginWrapper>
  );
};

Login.propTypes = {
  toggleNotification: PropTypes.func.isRequired,
};

export { Login };

export default withNotification(Login);
