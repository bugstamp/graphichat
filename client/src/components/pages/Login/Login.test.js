import React from 'react';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import Drawer from '@material-ui/core/Drawer';

import { Login } from './Login';
import LoginPresentation from './LoginPresentation';
import FullWidthSwipeableDrawer from '../../common/FullWidthSwipeableDrawer';
import { LoginWrapper } from './styled';

import storage from '../../../storage';

import mockedUseMediaQuery from '../../../__mocks__/@material-ui/core/useMediaQuery';
import { mountMockedProvider } from '../../../__mocks__/mockedProvider';
import { checkSessionExpirationMock } from '../../../__mocks__/mockedQueries';
import { tokens } from '../../../__mocks__/mockedQueryData';
import historyMock, { historyPushMock } from '../../../__mocks__/historyMock';
import useHookMock from '../../../__mocks__/useHookMock';
import { fixReactFacebookLoginBug } from '../../../__mocks__/testUtils';

const defaultMocks = [checkSessionExpirationMock];

describe('Login', () => {
  fixReactFacebookLoginBug();

  const toggleNotificationMock = jest.fn();
  const setStateMock = jest.fn();
  const useStateMock = jest.fn(() => [true, setStateMock]);

  const mountWrapper = (mocks = defaultMocks, options = {}) => mountMockedProvider(
    <Login history={historyMock} toggleNotification={toggleNotificationMock} />,
    mocks,
    options,
  );

  afterEach(() => {
    jest.clearAllMocks();
    storage.removeTokens();
  });

  test('should be mount without errors | intitial form state should be false', () => {
    const wrapper = mountWrapper();

    expect(wrapper.find('Login')).toHaveLength(1);
    expect(wrapper.find('LoginForm')).toHaveLength(0);
  });
  test('LoginWrapper should have container prop', () => {
    const wrapper = mountWrapper();
    const LoginWrapperTest = wrapper.find(LoginWrapper);

    expect(LoginWrapperTest).toHaveLength(1);
    expect(LoginWrapperTest.prop('container')).toBeTruthy();
  });
  test('toggleForm useState hook', async () => {
    const wrapper = mountWrapper();
    const LoginPresentationWrapper = wrapper.find(LoginPresentation);
    expect(LoginPresentationWrapper.prop('stopAnimation')).toBe(false);

    await act(async () => {
      LoginPresentationWrapper.prop('toggleForm')();

      await wait();
      wrapper.update();

      expect(wrapper.find(LoginPresentation).prop('stopAnimation')).toBe(true);
      expect(wrapper.find('LoginForm')).toHaveLength(1);
    });
  });
  test('toggleNotification should be called when initial sessionExpired state is true', () => {
    mountWrapper(null, { state: { sessionExpired: true } });

    expect(toggleNotificationMock).toBeCalledWith('Session time was expired');
  });
  test('Drawer should have correct props', () => {
    const wrapper = mountWrapper();
    const DrawerWrapper = wrapper.find(Drawer);

    expect(DrawerWrapper).toHaveLength(1);
    expect(DrawerWrapper.props()).toMatchObject({
      open: false,
      anchor: 'right',
    });
    expect(DrawerWrapper.prop('onClose')).toEqual(expect.any(Function));
  });
  test('Drawer should mount when useMediaQuery returns true', () => {
    const wrapper = mountWrapper();

    expect(wrapper.find(Drawer)).toHaveLength(1);
  });
  test('FullWidthSwipeableDrawer should mount when useMediaQuery returns false', () => {
    mockedUseMediaQuery.mockImplementation(() => false);
    const wrapper = mountWrapper();

    expect(wrapper.find(FullWidthSwipeableDrawer)).toHaveLength(1);
  });
  test('redirect to sign up page', async () => {
    useHookMock('useState', useStateMock);
    const wrapper = mountWrapper();

    await act(async () => {
      const redirectToSignUp = wrapper.find('LoginForm').prop('redirectToSignUp');

      expect(redirectToSignUp).toEqual(expect.any(Function));
      redirectToSignUp();
      expect(historyPushMock).toBeCalledWith('/reg');
    });
  });
  test('storage should save the tokens and history should redirect to main page when handleSuccess is invoked', async () => {
    useHookMock('useState', useStateMock);
    const wrapper = mountWrapper();
    const onSuccess = wrapper.find('LoginForm').prop('onSuccess');

    expect(onSuccess).toEqual(expect.any(Function));

    await act(async () => {
      onSuccess('signIn')({ signIn: tokens });

      expect(storage.getTokens()).toMatchObject(tokens);
      expect(historyPushMock).toBeCalledWith('/');
    });
  });
  test('handleError should invoke toggleNotification with error message', async () => {
    useHookMock('useState', useStateMock);
    const wrapper = mountWrapper();
    const message = 'Invalid field';
    const onError = wrapper.find('LoginForm').prop('onError');

    expect(onError).toEqual(expect.any(Function));

    await act(async () => {
      onError({ graphQLErrors: [{ message, data: { invalidField: null } }] });

      expect(toggleNotificationMock).toBeCalledWith(message);
    });
  });
});
