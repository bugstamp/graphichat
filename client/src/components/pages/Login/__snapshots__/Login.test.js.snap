// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login match to snapshot 1`] = `
<Fragment>
  <Login
    toggleNotification={[Function]}
  />
  <Notification
    message=""
    open={false}
    toggle={[Function]}
    type="error"
  />
</Fragment>
`;
